pennylane/workflow/execution.py

```diff
@@ -682,6 +537,9 @@ def cost_fn(params, x):
            [ 0.01983384, -0.97517033,  0.        ],
            [ 0.        ,  0.        , -0.95533649]])
     """
+    if not isinstance(device, qml.devices.Device):

```

Comment:
The facade is only applied if the user instantiates the device through `qml.device`. This check here is only a precautionary one in case the user creates the Legacy device independently through its constructor and passes it to the function.

Reply:
In that case, why not override `LegacyDevice.__new__` to always return a `LegacyDeviceFacade`?

Reply:
@albi3ro I actually think this might be quite efficient. What do you think?

Reply:
I see metaprogramming and Python black magic like `__new__` as a tool of last resort when we have no better options.  Yes, I have used it several times in the past, but that has also made me highly skeptical of it when we have different options.  

In this case, our other option is to just add a couple `if isinstance(device, LegacyDevice): device = LegacyDeviceFacade(device)`, which is pretty minimal in terms of code.  Any python developer can understand what it's doing.  I'm not concerned with the complexity and maintainence burden of those lines.